### PR TITLE
Rework NormalizedField to reference multiple objects

### PR DESCRIPTION
**Pull request by andimarek**
This changes NormalizedFields to contain a list of Object type names and not just one.

The main reason for this change change is to avoid an explosion of normalized Fields when you have Interfaces implemented by many Objects. This could lead to a huge amount of normalized fields, which had significant performant impact in certain cases.

It also aligns a NormalizedField more closely with what most people expect:

For example a schema

```graphql
type Query {
  pets: [Pet]
}
interface Pet {
  name: String
}
type Dog implements Pet {
  name: String
}
type Cat implements Pet {
  name: String
}
```
and this query:
```graphql
{pets{name}}
```

the normalized query would contain three fields: for each Object implementing Pet one  field: `Dog.name` and `Cat.name`. Now it is only one field referencing both: `[Dog, Cat].name`.

# Breaking Changes for DataFetchingFieldSelectionSet

As a consequence the API of `DataFetchingFieldSelectionSet` and `SelectionSet` changed.

`SelctionSet` was changed to reflect multiple objects now.

The API of `DataFetchingFieldSelectionSet` didn't change but the fully qualified names of a `SelectionSet` now contains a List of objects if there is more than one: e.g. `[Dog, Cat].name` instead of `Dog.name` and `Cat.name`.

This means some glob pattern may not work anymore as before. For example to match any `Cat` `name` field you need now a `*Cat*.name` instead of `Cat.name`.

### CODE REVIEW

**Comment by aleksandarsusnjar** [comment]
Early morning here... Just read your description. Definitely sounds like an improvement from my perspective, as I can (still) easily adjust my implementation and the present one is sort of dangerous for anyone who may ned up having a recursive schema. Will test very soon (but maybe not today exactly - will try).

Question though: Why do you need to "pre-normalize" at all? To me it seems that we, the `graphql-java` users have to handle the actual fetching and `graphql-java` has to get the values from our fetchers. Can't it decide at _that time_ what it needs based on actual/concrete type?

**Comment by bbakerman** [comment]
> Question though: Why do you need to "pre-normalize" at all? To me it seems that we, the graphql-java users have to handle the actual fetching and graphql-java has to get the values from our fetchers. Can't it decide at that time what it needs based on actual/concrete type?

Today the algorithm of graphql-java execution is based on the simple reference implementation

A breath first descent of the AST tree.  

However over the years we have learnt new ways of executing queries that may be more powerful for some use cases.  For example if we have a better "tree representation of the query (with types and arg values all pre rendered and resolved) then we could write an engine that did batching better.

At Atlassian we are writing a graphql gateway that doesnt execute queries per se but pulls them apart and feeds them onto underlying systems

We have learnt that a more optimised representation of the query allows for the engine to be easy to write and performs better.   This is where the original NormalisedQuery work came from.

We then learnt that the original implementation can explode out into millions of tree entries on certain schemas.  So this is revisiting those early decisions

 Imagine we could store that in a DB - we have a pre built query ready to execute - imagine we passed not query AST to an underlying graphql service, bit partial NormalisedQuery structures say.
 
 This is en evolution to a state where perhaps a graphql query is not parsed AST but in fact  a pre built NormalisedQuery with all the types and args resolved.

File: src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java [hunk 1]
```diff
@@ -253,7 +253,7 @@ private void executeImpl(ExecutionContext executionContext,
         Supplier<NormalizedField> normalizedFieldSupplier = getNormalizedField(executionContext, parameters, parameters::getExecutionStepInfo);
 
         GraphQLOutputType fieldType = fieldDef.getType();
-        DataFetchingFieldSelectionSet fieldCollector = DataFetchingFieldSelectionSetImpl.newCollector(fieldType, normalizedFieldSupplier);
+        DataFetchingFieldSelectionSet fieldCollector = DataFetchingFieldSelectionSetImpl.newCollector(executionContext.getGraphQLSchema(), fieldType, normalizedFieldSupplier);
 
         DataFetchingEnvironment environment = newDataFetchingEnvironment(executionContext)
                 .source(node.getSources())
```
Comment (bbakerman) [inline src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java:259]: We should kill Batch ES - its been deprecated for ages

File: src/main/java/graphql/normalized/NormalizedField.java [hunk 2]
```diff
@@ -110,40 +165,30 @@ public NormalizedField transform(Consumer<Builder> builderConsumer) {
...
-        NormalizedField cur = this;
-        while (cur != null) {
-            if (result.length() > 0) {
-                result.insert(0, "/");
-            }
-            result.insert(0, cur.print());
-            cur = cur.getParent();
+    public String objectTypeNamesToString() {
+        if (objectTypeNames.size() == 1) {
+            return objectTypeNames.iterator().next();
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/NormalizedField.java:188]: list.get(0) is quicker - index look up  - no allocation of iterator

```
        return (E) elementData[index];

```

File: src/main/java/graphql/normalized/NormalizedField.java [hunk 3]
```diff
@@ -1,52 +1,107 @@
...
 public class NormalizedField {
     private final String alias;
     private final Map<String, NormalizedInputValue> normalizedArguments;
-    private final Map<String, Object> arguments;
-    private final GraphQLObjectType objectType;
-    private final GraphQLFieldDefinition fieldDefinition;
+    private final Map<String, Object> resolvedArguments;
+    private final ImmutableList<Argument> astArguments;
+    private final Set<String> objectTypeNames;
+    private final String fieldName;
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/NormalizedField.java:37]: Do we use ImmutableSets here

File: src/main/java/graphql/normalized/NormalizedField.java [hunk 4]
```diff
@@ -1,52 +1,107 @@
...
+        return getOneFieldDefinition(schema).getType();
+    }
+
+    public GraphQLFieldDefinition getOneFieldDefinition(GraphQLSchema schema) {
+        GraphQLFieldDefinition fieldDefinition;
+        GraphQLFieldDefinition introspectionField = resolveIntrospectionField(fieldName, schema);
+        if (introspectionField != null) {
+            return introspectionField;
+        }
+        GraphQLObjectType type = (GraphQLObjectType) assertNotNull(schema.getType(objectTypeNames.iterator().next()));
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/NormalizedField.java:72]: `get(0)` is quicker

File: src/main/java/graphql/normalized/NormalizedField.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/normalized/NormalizedField.java:188]: it is a Set, so no get(0)

File: src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java [hunk 5]
```diff
@@ -113,34 +144,301 @@ private void fixUpParentReference(NormalizedField rootNormalizedField) {
...
+        if (value1 == null && value2 == null) {
+            return true;
+        }
+        if (value1 == null) {
+            return false;
+        }
+        if (value2 == null) {
+            return false;
+        }
+        return new AstComparator().isEqual(value1, value2);
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java:398]: could this code go in AstComparator() -

Also does AstComparator need to be an object allocation?  Why not a static method

File: src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java [hunk 6]
```diff
@@ -113,34 +144,301 @@ private void fixUpParentReference(NormalizedField rootNormalizedField) {
...
+            if (fieldName.equals(schema.getIntrospectionSchemaFieldDefinition().getName())) {
+                fieldDefinition = schema.getIntrospectionSchemaFieldDefinition();
+            } else if (fieldName.equals(schema.getIntrospectionTypeFieldDefinition().getName())) {
+                fieldDefinition = schema.getIntrospectionTypeFieldDefinition();
+            } else {
+                fieldDefinition = assertNotNull(objectType.getField(fieldName), () -> String.format("no field %s found for type %s", fieldName, objectType));
+            }
+        }
+        return fieldDefinition;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java:416]: `graphql.introspection.Introspection#getFieldDef` does this already

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java [hunk 7]
```diff
@@ -311,18 +316,28 @@ public String getFullyQualifiedName() {
...
+        }
+
+        @Override
+        public GraphQLOutputType getType() {
+            return normalizedField.getType(schema);
+        }
+
+        @Override
+        public List<GraphQLObjectType> getObjectTypes() {
+            return (List) this.schema.getTypes(normalizedField.getObjectTypeNames());
```
Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:330]: cast??

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java [hunk 8]
```diff
@@ -311,18 +316,28 @@ public String getFullyQualifiedName() {
...
         }
 
         @Override
-        public GraphQLFieldDefinition getFieldDefinition() {
-            return normalizedField.getFieldDefinition();
+        public List<String> getObjectTypeNames() {
+            return new ArrayList<>(normalizedField.getObjectTypeNames());
```
Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:335]: ImmutableList.of() - no allocation if its immutable I think

Comment (bbakerman) [reply src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:335]:
```
    public static <E> ImmutableList<E> copyOf(Collection<? extends E> elements) {
        if (elements instanceof ImmutableCollection) {
            ImmutableList<E> list = ((ImmutableCollection)elements).asList();
            return list.isPartialView() ? asImmutableList(list.toArray()) : list;
        } else {
            return construct(elements.toArray());
        }
    }

```

Yup no allocation on immutable collections

Comment (bbakerman) [reply src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:335]: well some but smarter

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java [hunk 9]
```diff
@@ -311,18 +316,28 @@ public String getFullyQualifiedName() {
...
         }
 
         @Override
         public Map<String, Object> getArguments() {
-            return normalizedField.getArguments();
+            return normalizedField.getResolvedArguments();
```
Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:340]: Immutable ??

File: src/main/java/graphql/schema/GraphQLSchema.java [hunk 10]
```diff
@@ -250,10 +252,24 @@ public GraphQLObjectType getIntrospectionSchemaType() {
...
      * @return the type
      */
-    public GraphQLType getType(String typeName) {
+    public @Nullable GraphQLType getType(@NotNull String typeName) {
         return typeMap.get(typeName);
     }
 
+    /**
+     * All types with the provided names.
+     * throws Exception when a type name could not be resolved
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLSchema.java:261]: Throws WHAT exception ??  please document

File: src/test/java/benchmark/NQBenchmark2.java [hunk 11]
```diff
@@ -62,15 +71,48 @@ private String readFromClasspath(String file) throws IOException {
...
+        Traverser<NormalizedField> traverser = Traverser.depthFirst(NormalizedField::getChildren);
+        traverser.traverse(queryExecutionTree.getTopLevelFields(), new TraverserVisitorStub<NormalizedField>() {
+            @Override
+            public TraversalControl enter(TraverserContext<NormalizedField> context) {
+                NormalizedField queryExecutionField = context.thisNode();
+                result.add(queryExecutionField.printDetails());
+                return TraversalControl.CONTINUE;
+            }
+        });
+        return result;
```
Comment (bbakerman) [inline src/test/java/benchmark/NQBenchmark2.java:116]: Should something like this go on NormalisedQuery itself??

**Review by bbakerman — APPROVED** [review]
I think we should do one more pass on Immutable structures here.  This will improve the memory profile of this even more

Also dont rely on the iterator.next trick.  (Ahh is it a Set ....maybe you have to)

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java (hunk 8 above)
Comment (andimarek) [reply src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:335]: unfortunately object types names is a mutable Set (on purpose)

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java (hunk 7 above)
Comment (andimarek) [reply src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:330]: fixed

File: src/main/java/graphql/normalized/NormalizedField.java (hunk 4 above)
Comment (andimarek) [reply src/main/java/graphql/normalized/NormalizedField.java:72]: it si a Set :-(

File: src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java:259]: will do in a separate PR

File: src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java (hunk 6 above)
Comment (andimarek) [reply src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java]: thanks!

File: src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java (hunk 9 above)
Comment (andimarek) [reply src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java:340]: changed to returning an in immutable Map

File: src/test/java/benchmark/NQBenchmark2.java (hunk 11 above)
Comment (andimarek) [reply src/test/java/benchmark/NQBenchmark2.java:116]: moved

Comment (andimarek) [reply src/test/java/benchmark/NQBenchmark2.java:116]: maybe yes ... but I will not do it now.

File: src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java (hunk 5 above)
Comment (andimarek) [reply src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java]: great point ... moved

**Comment by aleksandarsusnjar** [comment]
> At Atlassian we are writing a graphql gateway that doesnt execute queries per se but pulls them apart and feeds them onto underlying systems
> 
> We have learnt that a more optimised representation of the query allows for the engine to be easy to write and performs better. This is where the original NormalisedQuery work came from.
> 
> We then learnt that the original implementation can explode out into millions of tree entries on certain schemas. So this is revisiting those early decisions
> 
> Imagine we could store that in a DB - we have a pre built query ready to execute - imagine we passed not query AST to an underlying graphql service, bit partial NormalisedQuery structures say.

What I am doing with `graphql-java`, and I imagine many are to some extent, is exposing stuff _we_ store in the DB. I have a model we wish to expose. My code is doing the refactoring of the GraphQL queries into "somewhat more" efficient  underlying queries based on our assumptions and statistics. Note that the GraphQL data model isn't the same as the storage model so the GraphQL "engine" can't really know what is the most efficient way of executing those (I am not talking about aggregators from multiple sources). By trying to be "helpful" it is taking away the original query model that is may be hiding the actual user's intent as well. What I am doing now is almost literally "reverse engineering" that normalization - trying to get what did the author of the query really say. This may sound unimportant but that provides the author with some way to influence the optimizer and nudge it a different way, intentionally - as all the optimization intelligence based on statistics can and does break down here-and-there.

Note that for every exposed type I have my own "mapping metadata". I also keep a graph of all types with links to their immediate/direct sub- and super-types and also have fast means to find all (not only immediate/direct) sub- and super-types as well. I can check type relationships, property declarations and inheritance, etc. very quickly. I can tell you exactly in which types a property exists in nanonseconds. Sure, this takes some memory but only "once" (per schema).

**Pull request merged**